keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,652 @@
+Users or controllers can add a string to `Liens` to ask to protect a resource.
+A newly introduced validating admission webhook will reject deletion requests for resources with non-empty `Liens`.
+
+`Liens` is defined as a slice of strings, like `Finalizers`.
+Multiple users or controllers can add their `Liens` for their own purpose, and they can remove their own `Liens` when it is no longer needed.
+Deletion requests of the resource are blocked until the last `Liens` for the resource is removed
+(Just to be clear, the difference between `Liens` and `Finalizers` is that `Liens` blocks the deletion request itself, while `Finalizers` blocks the deletion to be completed).
+
+To make multiple users and controllers share the `Liens` field properly, they should include their identity to the `Liens` string and they should only delete their own `Liens` strings.
+
+A PoC implementation for in-use protection can be found, [here](https://github.com/mkimuram/kubernetes/commits/lien).
+Also, how it can be consumed by Secret protection controller can be found, [here](TBD).
+
+### API Changes
+
+`Liens` is added to `ObjectMeta`.
+
+```
+Liens []string
+```
+
+### Other Design Considerations
+
+#### Opt-out Protection
+
+In some cases, users may need to opt-out protection to force delete the resrouce.
+However, controllers may regularly add `Liens` to a resource, therefore, just removing `Liens` before force delete won't be enough.
+To avoid such a situation, below opt-out mechanism is provided.
+
+- If `k8s.io/skip-lien` annotation is added to the resource, the resource is regarded as it has empty `Liens`. Therefore, the resource is deleted regardless of its `Liens` field.
+
+#### Block Adding Additional `Liens` while Deleting
+
+`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
+Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
+However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
+To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by validating admission webhook for update.
+
+#### Cross Namespace Dependency Consideration
+
+`Liens` can be added only by users or controllers that have create/update rights to resources.
+Therefore, this mechanism itself doesn't have issues for cross namespace dependency.
+However, when it is used with controllers that have access across namespaces,
+they can easily introduce a cross namespace dependency.
+For example, adding `Liens` to resource X in namespace x, because resource Y exists in namespace y.
+This may allow malicious users to block the resources in other namespace from being deleted.
```

Comment:
I don't buy this, if you have update permission in your namespace you can delete the thing anyway.